File: delivery/php-sdk/models/sender/api/idelivery_sender.php
<?php
/**
 * @Author Hanqiang
 * @Date 2025/12/9
 * @PackageName: models/sender/api
 * @ClassName: DeliverySender
 * @Description: 麦芽田配送开放平台SDK - 配送服务发送端接口和实现
 * Sender代表三方配送服务商主动调用麦芽田开放平台的接口
 * @Version 1.0
 */

namespace Maiyatian\Delivery\SDK\models\sender\api;

use Maiyatian\Delivery\SDK\client\Config;
use Maiyatian\Delivery\SDK\client\HTTPClientManager;
use Maiyatian\Delivery\SDK\models\sender\entity\auth\AccessTokenRequest;
use Maiyatian\Delivery\SDK\models\sender\entity\auth\RefreshTokenRequest;
use Maiyatian\Delivery\SDK\models\sender\entity\delivery\DeliveryChangeRequest;
use Maiyatian\Delivery\SDK\models\sender\entity\express\LocationChangeRequest;
use Maiyatian\Delivery\SDK\models\types\Constants;

/**
 * IDeliverySender 配送服务发送端接口定义
 *
 * 接口说明:
 *   - Sender接口用于三方配送服务商主动向麦芽田平台推送数据和请求服务
 *   - 包括：配送状态同步、快递轨迹回传、授权相关接口
 *
 * 使用场景:
 *   - 三方配送服务商需要向麦芽田平台推送配送状态时
 *   - 三方配送服务商需要推送快递轨迹信息时
 *   - 三方配送服务商需要获取或刷新访问令牌时
 */
interface IDeliverySender
{
    /**
     * 配送状态同步(必接)
     *
     * 功能: 同步订单配送状态操作
     * 说明: 三方配送服务商主动推送订单配送状态到麦芽田平台
     * command: delivery_change
     * 必接: 是
     *
     * @param string $token 授权令牌
     * @param DeliveryChangeRequest $data 配送状态同步请求数据
     * @return HTTPClientManager\HTTPResponse
     */
    public function deliveryChange(string $token, DeliveryChangeRequest $data);

    /**
     * 快递轨迹回传(必接)
     *
     * 功能: 推送快递轨迹信息
     * 说明: 三方配送服务商主动推送快递轨迹信息到麦芽田平台
     * command: location_change
     * 必接: 是
     *
     * @param string $token 授权令牌
     * @param LocationChangeRequest $data 快递轨迹回传请求数据
     * @return HTTPClientManager\HTTPResponse
     */
    public function locationChange(string $token, LocationChangeRequest $data);

    /**
     * 授权码code换取访问令牌Token(必接)
     *
     * 功能: 用授权码换取访问令牌
     * 说明: 用授权码code换取长时令牌refreshToken以及访问令牌accessToken
     * command: access_token
     * 必接: 是
     *
     * @param AccessTokenRequest $data 授权码获取access_token请求数据
     * @return HTTPClientManager\HTTPResponse
     */
    public function accessToken(AccessTokenRequest $data);

    /**
     * 刷新访问令牌Token(必接)
     *
     * 功能: 刷新过期的访问令牌
     * 说明: 当token过期时，使用refresh_token重新获取新的token
     * command: refresh_token
     * 必接: 是
     *
     * @param string $token 授权令牌
     * @param RefreshTokenRequest $data 刷新访问令牌请求数据
     * @return HTTPClientManager\HTTPResponse
     */
    public function refreshToken(string $token, RefreshTokenRequest $data);
}

/**
 * DeliverySender 配送服务发送端实现
 */
class DeliverySender implements IDeliverySender
{
    /**
     * @var HTTPClientManager
     */
    protected $clientManager;

    /**
     * DeliverySender constructor.
     *
     * @param HTTPClientManager $clientManager
     */
    public function __construct(HTTPClientManager $clientManager)
    {
        $this->clientManager = $clientManager;
    }

    /**
     * 获取HTTP客户端管理器
     *
     * @return HTTPClientManager
     */
    public function getClientManager(): HTTPClientManager
    {
        return $this->clientManager;
    }

    /**
     * 设置HTTP客户端管理器
     *
     * @param HTTPClientManager $clientManager
     */
    public function setClientManager(HTTPClientManager $clientManager): void
    {
        $this->clientManager = $clientManager;
    }

    /**
     * {@inheritdoc}
     */
    public function deliveryChange(string $token, DeliveryChangeRequest $data)
    {
        return $this->clientManager->post(
            Constants::COMMAND_DELIVERY_CHANGE,
            $data,
            $token
        );
    }

    /**
     * {@inheritdoc}
     */
    public function locationChange(string $token, LocationChangeRequest $data)
    {
        return $this->clientManager->post(
            Constants::COMMAND_LOCATION_CHANGE,
            $data,
            $token
        );
    }

    /**
     * {@inheritdoc}
     */
    public function accessToken(AccessTokenRequest $data)
    {
        return $this->clientManager->post(
            Constants::COMMAND_ACCESS_TOKEN,
            $data
        );
    }

    /**
     * {@inheritdoc}
     */
    public function refreshToken(string $token, RefreshTokenRequest $data)
    {
        return $this->clientManager->post(
            Constants::COMMAND_REFRESH_TOKEN,
            $data,
            $token
        );
    }
}

/**
 * DeliverySenderFactory 配送服务发送端工厂类
 */
class DeliverySenderFactory
{
    /**
     * 创建配送服务发送端实例
     *
     * 功能说明:
     *   - 创建用于三方配送服务商主动调用麦芽田平台的客户端实例
     *   - 初始化HTTP客户端管理器
     *
     * @param Config $config HTTP客户端配置
     * @return IDeliverySender
     */
    public static function create(Config $config): IDeliverySender
    {
        $clientManager = new HTTPClientManager($config);
        return new DeliverySender($clientManager);
    }
}

File: delivery/java-sdk/src/main/java/com/maiyatian/delivery/sdk/models/sender/api/IDeliverySender.java
/*
 * 麦芽田配送开放平台 Java SDK
 * Copyright (c) 2025 麦芽田开放平台
 * MIT License
 */

package com.maiyatian.delivery.sdk.models.sender.api;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.maiyatian.delivery.sdk.models.sender.entity.auth.AccessTokenEntity;
import com.maiyatian.delivery.sdk.models.sender.entity.auth.RefreshTokenEntity;
import com.maiyatian.delivery.sdk.models.sender.entity.delivery.DeliveryChangeEntity;
import com.maiyatian.delivery.sdk.models.sender.entity.express.LocationChangeEntity;
import com.maiyatian.delivery.sdk.models.types.Response;

import java.io.IOException;
import java.util.Map;

/**
 * 配送发送接口
 * <p>
 * 定义三方服务主动推送数据到麦芽田开放平台的所有接口
 * </p>
 * 
 * <h3>接口说明：</h3>
 * 这些接口由三方服务商主动调用，向麦芽田开放平台推送配送状态变更、位置变更等信息
 * 
 * <h3>接口分类：</h3>
 * <ul>
 * <li><strong>授权管理</strong>：获取访问令牌、刷新令牌</li>
 * <li><strong>配送变更</strong>：配送状态变更推送</li>
 * <li><strong>位置变更</strong>：骑手位置变更推送</li>
 * </ul>
 * 
 * @author SDK Generator
 * @version 1.0.0
 * @since 1.0.0
 */
public interface IDeliverySender extends AutoCloseable {
    
    /**
     * 关闭资源
     * 
     * @throws IOException 如果关闭资源时发生错误
     */
    @Override
    void close() throws IOException;

    // ==================== 授权管理接口 ====================

    /**
     * 获取访问令牌接口
     * <p>
     * 三方服务使用授权码code向麦芽田换取访问令牌
     * </p>
     * 
     * <h3>业务流程：</h3>
     * 商户授权完成 → 三方获得code → 调用此接口向麦芽田换取token → 麦芽田返回访问令牌
     * 
     * <h3>接口路径：</h3>
     * POST /v1/delivery/access_token
     * 
     * @param token 商户授权令牌
     * @param data 授权请求数据
     * @param headers 自定义请求头（可选）
     * @return 访问令牌响应
     * @throws IOException 网络请求失败时
     * @throws JsonProcessingException JSON序列化失败时
     */
    AccessTokenEntity.Resp accessToken(String token, AccessTokenEntity.Req data, Map<String, String> headers)
            throws IOException, JsonProcessingException;

    /**
     * 获取访问令牌接口（简化版本）
     */
    AccessTokenEntity.Resp accessToken(String token, AccessTokenEntity.Req data)
            throws IOException, JsonProcessingException;

    /**
     * 刷新访问令牌接口
     * <p>
     * 三方服务使用刷新令牌向麦芽田重新获取新的访问令牌
     * </p>
     * 
     * <h3>业务流程：</h3>
     * token即将过期 → 调用此接口向麦芽田刷新 → 麦芽田返回新的token和refresh_token
     * 
     * <h3>接口路径：</h3>
     * POST /v1/delivery/refresh_token
     * 
     * @param token 商户授权令牌
     * @param data 刷新请求数据
     * @param headers 自定义请求头（可选）
     * @return 刷新令牌响应
     * @throws IOException 网络请求失败时
     * @throws JsonProcessingException JSON序列化失败时
     */
    RefreshTokenEntity.Resp refreshToken(String token, RefreshTokenEntity.Req data, Map<String, String> headers)
            throws IOException, JsonProcessingException;

    /**
     * 刷新访问令牌接口（简化版本）
     */
    RefreshTokenEntity.Resp refreshToken(String token, RefreshTokenEntity.Req data)
            throws IOException, JsonProcessingException;

    // ==================== 配送变更接口 ====================

    /**
     * 配送状态变更推送接口
     * <p>
     * 三方服务在配送状态发生变更时，主动推送变更信息给麦芽田
     * </p>
     * 
     * <h3>业务流程：</h3>
     * 配送状态变更 → 三方调用此接口推送给麦芽田 → 麦芽田更新配送状态
     * 
     * <h3>接口路径：</h3>
     * POST /v1/delivery/delivery_change
     * 
     * @param token 商户授权令牌
     * @param data 配送变更数据
     * @param headers 自定义请求头（可选）
     * @return API响应
     * @throws IOException 网络请求失败时
     * @throws JsonProcessingException JSON序列化失败时
     */
    Response deliveryChange(String token, DeliveryChangeEntity.Req data, Map<String, String> headers)
            throws IOException, JsonProcessingException;

    /**
     * 配送状态变更推送接口（简化版本）
     */
    Response deliveryChange(String token, DeliveryChangeEntity.Req data)
            throws IOException, JsonProcessingException;

    // ==================== 位置变更接口 ====================

    /**
     * 骑手位置变更推送接口
     * <p>
     * 三方服务在骑手位置发生变更时，主动推送位置信息给麦芽田
     * </p>
     * 
     * <h3>业务流程：</h3>
     * 骑手位置变更 → 三方调用此接口推送给麦芽田 → 麦芽田更新骑手位置
     * 
     * <h3>接口路径：</h3>
     * POST /v1/delivery/location_change
     * 
     * @param token 商户授权令牌
     * @param data 位置变更数据
     * @param headers 自定义请求头（可选）
     * @return API响应
     * @throws IOException 网络请求失败时
     * @throws JsonProcessingException JSON序列化失败时
     */
    Response locationChange(String token, LocationChangeEntity.Req data, Map<String, String> headers)
            throws IOException, JsonProcessingException;

    /**
     * 骑手位置变更推送接口（简化版本）
     */
    Response locationChange(String token, LocationChangeEntity.Req data)
            throws IOException, JsonProcessingException;

    // ==================== 工具方法 ====================

    /**
     * 执行原始HTTP请求
     * 
     * @param method HTTP方法
     * @param path API路径
     * @param data 请求数据
     * @param token 授权令牌
     * @param headers 请求头
     * @return API响应
     * @throws IOException 网络请求失败时
     * @throws JsonProcessingException JSON处理失败时
     */
    Response request(String method, String path, Object data, String token, Map<String, String> headers)
            throws IOException, JsonProcessingException;
}
